druid/src/menu/mod.rs

```diff
@@ -0,0 +1,762 @@
+        MenuBuildCtx {
+            current: if popup {
+                PlatformMenu::new_for_popup()
+            } else {
+                PlatformMenu::new()
+            },
+        }
+    }
+
+    pub fn with_submenu(&mut self, text: &str, enabled: bool, f: impl FnOnce(&mut MenuBuildCtx)) {
+        let mut child = MenuBuildCtx::new(false);
+        f(&mut child);
+        self.current.add_dropdown(child.current, text, enabled);
+    }
+
+    pub fn add_item(
+        &mut self,
+        id: u32,
+        text: &str,
+        key: Option<&HotKey>,
+        enabled: bool,
+        selected: bool,
+    ) {
+        self.current.add_item(id, text, key, enabled, selected);
+    }
+
+    pub fn add_separator(&mut self) {
+        self.current.add_separator();
+    }
+}
+
+impl<'a> MenuEventCtx<'a> {
+    /// Submit a [`Command`] to be handled by the main widget tree.
+    ///
+    /// See [`EventCtx::submit_command`] for more information.
+    ///
+    /// [`Command`]: crate::Command
+    /// [`EventCtx::submit_command`]: crate::EventCtx::submit_command
+    pub fn submit_command(&mut self, cmd: impl Into<Command>) {
+        self.queue.push_back(cmd.into());
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum MenuUpdate {
+    /// The structure of the current menu is ok, but some elements need to be refreshed (e.g.
+    /// changing their text, whether they are enabled, etc.)
+    NeedsRefresh,
+    /// The structure of the menu has changed; we need to rebuild from scratch.
+    NeedsRebuild,
+    /// No need to rebuild anything.
+    UpToDate,
+}
+
+impl MenuUpdate {
+    fn combine(self, other: MenuUpdate) -> MenuUpdate {
+        use MenuUpdate::*;
+        match (self, other) {
+            (NeedsRebuild, _) | (_, NeedsRebuild) => NeedsRebuild,
+            (NeedsRefresh, _) | (_, NeedsRefresh) => NeedsRefresh,
+            _ => UpToDate,
+        }
+    }
+}
+
+/// This is the trait that enables recursive visiting of all menu entries. It isn't publically
+/// visible (the publically visible analogue of this is `Into<MenuEntry<T>>`).
+trait MenuTrait<T> {
+    /// Called when a menu item is activated.
+    ///
+    /// `id` is the id of the entry that got activated. If this is different from your id, you are
+    /// responsible for routing the activation to your child items.
+    fn activate(&mut self, ctx: &mut MenuEventCtx, id: MenuItemId, data: &mut T, env: &Env);
+
+    /// Called when the data is changed.
+    fn update(&mut self, old_data: &T, data: &T, env: &Env) -> MenuUpdate;
+
+    /// Called to refresh the menu.
+    fn refresh(&mut self, ctx: &mut MenuBuildCtx, data: &T, env: &Env);
+}
+
+/// A wrapper for a menu item (or submenu) to give it access to a part of its parent data.
+///
+/// This is the menu analogue of [`LensWrap`]. You will usually create it with [`Menu::lens`] or
+/// [`MenuItem::lens`] instead of using this struct directly.
+///
+/// [`LensWrap`]: crate::widget::LensWrap
+pub struct MenuLensWrap<L, U> {
```

Comment:
we could consider also folding some of the `WidgetPod` functionality into here; in particular we could stash the last seen data and env, and then stop recursing if they're unchanged?